### PR TITLE
Remove dependency on Sigil 4.7.0 / Sigil-vNext 4.8.41

### PR DESCRIPTION
**Pull request by zacharycmontoya**
### WIP Status
- Perf test

### Summary
This PR removes usages of `Sigil 4.7.0` and `Sigil-vNext 4.8.41`, removing `Sigil.dll` dependency on all target frameworks. After these changes:
1. Accessing fields and properties on application objects is done through `PropertyFetcher` and `FieldFetcher` types
2. Calling methods on application objects is done through a delegate created by `System.Linq.Expressions`
3. Calling the original call target for automatic instrumentation is done through a delegate created by `System.Reflection.Emit.ILGenerator`

### Changes
- Remove `Sigil` NuGet packages from `Datadog.Trace.ClrProfiler.Managed.csproj` and add direct NuGet package references to `System.Reflection.Emit` NuGet packages that were previously transitively referenced for the `netstandard2.0` build target
- Remove old `Datadog.Trace.ClrProfiler.DynamicMethodBuilder` which has no internal uses and cannot be statically referenced by application developers
- Create a new `FieldFetcher` class which is based on the `PropertyFetcher` class to perform fast field lookups. Make both classes implement the `IMemberFetcher` interface for consistency in API
- Fix `ExecuteReaderAsync` in the `SqlCommandIntegration` and `NpgsqlCommandIntegration` classes that began failing after removing Sigil. The return types were specified by the delegate to be `Task<DbDataReader>` even though the public interfaces were specified to return `Task<SqlDataReader>` and `Task<NpqgsqlDataReader>`
- `Datadog.Trace.ClrProfiler.Emit.ObjectExtensions`
  - Remove `CreatePropertyDelegate` / `CreateFieldDelegate` so all dynamic property fetches are done through the `PropertyFetcher` / `FieldFetcher` classes
  - Rename `PropertyFetcherCacheKey` to `MemberFetcherCacheKey` to better describe that property, field, and method keys are all being generated by this type
  - Store static field `System.Reflection.Emit.ModuleBuilder Module` to hold single module in which `MethodBuilder` dynamic methods will be emitted
- `Datadog.Trace.ClrProfiler.Emit.MethodBuilder`
  - Replace all usages of `Sigil.Emit<T>` with `System.Reflection.Emit.ILGenerator` (this is the last remaining use of `System.Reflection.Emit.ILGenerator`)
- `Datadog.Trace.ClrProfiler.Emit.DynamicMethodBuilder'1`
  - Replace all usages of `Sigil.Emit<T>` with `System.Linq.Expressions`

### Benchmarks
```
|                                       Method |     Toolchain |     Mean |     Error |    StdDev | Ratio | RatioSD |
|--------------------------------------------- |-------------- |---------:|----------:|----------:|------:|--------:|
|                           GetField_Old_Sigil |        net472 | 4.924 ns | 0.1279 ns | 0.1368 ns |  1.32 |    0.07 |
|                    GetField_New_FieldFetcher |        net472 | 8.365 ns | 0.2275 ns | 0.6417 ns |  1.29 |    0.12 |
|                                              |               |          |           |           |       |         |
|                           GetField_Old_Sigil | netcoreapp3.1 | 3.734 ns | 0.1029 ns | 0.1661 ns |  1.00 |    0.00 |
|                    GetField_New_FieldFetcher | netcoreapp3.1 | 6.527 ns | 0.1563 ns | 0.4383 ns |  1.00 |    0.00 |
|                                              |               |          |           |           |       |         |
|           CreateMethodCallDelegate_Old_Sigil |        net472 | 6.018 ns | 0.1505 ns | 0.2009 ns |  1.22 |    0.08 |
| CreateMethodCallDelegate_New_LinqExpressions |        net472 | 9.388 ns | 0.2092 ns | 0.4030 ns |  3.51 |    0.32 |
|                                              |               |          |           |           |       |         |
|           CreateMethodCallDelegate_Old_Sigil | netcoreapp3.1 | 4.778 ns | 0.1262 ns | 0.2797 ns |  1.00 |    0.00 |
| CreateMethodCallDelegate_New_LinqExpressions | netcoreapp3.1 | 2.690 ns | 0.0897 ns | 0.2266 ns |  1.00 |    0.00 |
|                                              |               |          |           |           |       |         |
|          MethodBuilderEmitDelegate_Old_Sigil |        net472 | 3.593 ns | 0.1024 ns | 0.2921 ns |  0.92 |    0.07 |
|         MethodBuilderEmitDelegate_New_ILEmit |        net472 | 2.099 ns | 0.0440 ns | 0.0368 ns |  0.99 |    0.02 |
|                                              |               |          |           |           |       |         |
|          MethodBuilderEmitDelegate_Old_Sigil | netcoreapp3.1 | 3.754 ns | 0.1008 ns | 0.1738 ns |  1.00 |    0.00 |
|         MethodBuilderEmitDelegate_New_ILEmit | netcoreapp3.1 | 2.118 ns | 0.0303 ns | 0.0269 ns |  1.00 |    0.00 |
```
@DataDog/apm-dotnet

### CODE REVIEW

**Comment by lucaspimentel** [comment]
If I'm reading the benchmarks correctly, this is _roughly_:

Method | .NET Framework | .NET Core
--|--|--
`GetField` | 75% slower | 75% slower
`CreateMethodCallDelegate` | 50% slower | 50% faster
`MethodBuilderEmitDelegate` | 40% faster | 40% faster

- `GetField` now takes twice as long. How often do we use this? Is this change acceptable or should we try emitting IL instead?
- `CreateMethodCallDelegate` is slower in .NET Framework, faster in .NET Core 🤷🏽‍♂️
- `MethodBuilderEmitDelegate` is always faster. Makes sense since both versions emit IL and the new one avoids Sigil's wrapper and IL validation. 👍🏽

**Comment by zacharycmontoya** [comment]
I think it's misleading to look at these as percentages since 75% slower means ~3ns. These are also cases where the called methods are returning methods instantaneously so any larger amount of work should make this negligible. With the current results, my opinion is that these are acceptable unless our larger E2E reliability environment indicates a larger regression.

**Comment by lucaspimentel** [comment]
Makes sense, thanks!

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNet/NpgsqlCommandIntegration.cs [hunk 1]
```diff
@@ -217,11 +217,35 @@ public static class NpgsqlCommandIntegration
                 throw;
             }
 
+            return AsyncHelper.InvokeGenericTaskDelegate(
+                owningType: command.GetType(),
+                taskResultType: npgsqlDataReaderType,
+                nameOfIntegrationMethod: nameof(ExecuteReaderAsyncInternal),
+                integrationType: typeof(NpgsqlCommandIntegration),
+                (DbCommand)command,
+                cancellationToken,
```
Comment (kevingosse) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNet/NpgsqlCommandIntegration.cs:226]:
```suggestion
                boxedCancellationToken,
```

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNet/NpgsqlCommandIntegration.cs [hunk 2]
```diff
@@ -303,11 +327,38 @@ public static class NpgsqlCommandIntegration
...
             }
 
+            return AsyncHelper.InvokeGenericTaskDelegate(
+                owningType: command.GetType(),
+                taskResultType: npgsqlDataReaderType,
+                nameOfIntegrationMethod: nameof(ExecuteReaderAsyncWithBehaviorAndCancellationInternal),
+                integrationType: typeof(NpgsqlCommandIntegration),
+                (DbCommand)command,
+                commandBehavior,
+                cancellationToken,
```
Comment (kevingosse) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNet/NpgsqlCommandIntegration.cs:337]:
```suggestion
                boxedCancellationToken,
```

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNet/SqlCommandIntegration.cs [hunk 3]
```diff
@@ -218,11 +219,38 @@ public static class SqlCommandIntegration
...
             }
 
+            return AsyncHelper.InvokeGenericTaskDelegate(
+                owningType: command.GetType(),
+                taskResultType: sqlDataReaderType,
+                nameOfIntegrationMethod: nameof(ExecuteReaderAsyncInternal),
+                integrationType: typeof(SqlCommandIntegration),
+                (DbCommand)command,
+                commandBehavior,
+                cancellationToken,
```
Comment (kevingosse) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNet/SqlCommandIntegration.cs:229]:
```suggestion
                boxedCancellationToken,
```

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNet/SqlCommandIntegration.cs [hunk 4]
```diff
@@ -218,11 +219,38 @@ public static class SqlCommandIntegration
                 throw;
             }
 
+            return AsyncHelper.InvokeGenericTaskDelegate(
```
Comment (kevingosse) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNet/SqlCommandIntegration.cs:222]: You could instead do like HttpMessageHandlerIntegration:

- Make the invoked async method return Task<object>
- Invoke it directly and use the TaskExtensions.Cast helper to convert the return type
- In the invoked async method, cast the return value of instrumentedMethod to Task to await it, then use `.GetProperty("Result").Value` to fetch the result

If was faster for HttpMessageHandlerIntegration, so I expect the same here. But we can do that in another PR if you prefer

**Comment by zacharycmontoya** [comment]
Closing in favor of #993 which makes the minimum changes needed without introducing additional risk at this time.

**Pull request closed without merging**